https://adventofcode.com/2021/day/7

<h2>--- Day 7: The Treachery of Whales ---</h2><p>A giant <a href="https://en.wikipedia.org/wiki/Sperm_whale" target="_blank">whale</a> has decided your submarine is its next meal, and it's much faster than you are. There's nowhere to run!</p>
<p>Suddenly, a swarm of crabs (each in its own tiny submarine - it's too deep for them otherwise) zooms in to rescue you! They seem to be preparing to blast a hole in the ocean floor; sensors indicate a <em>massive underground cave system</em> just beyond where they're aiming!</p>
<p>The crab submarines all need to be aligned before they'll have enough power to blast a large enough hole for your submarine to get through. However, it doesn't look like they'll be aligned before the whale catches you! Maybe you can help?</p>
<p>There's one major catch - crab submarines can only move horizontally.</p>
<p>You quickly make a list of <em>the horizontal position of each crab</em> (your puzzle input). Crab submarines have limited fuel, so you need to find a way to make all of their horizontal positions match while requiring them to spend as little fuel as possible.</p>
<p>For example, consider the following horizontal positions:</p>
<pre><code>16,1,2,0,4,2,7,1,2,14</code></pre>
<p>This means there's a crab with horizontal position <code>16</code>, a crab with horizontal position <code>1</code>, and so on.</p>
<p>Each change of 1 step in horizontal position of a single crab costs 1 fuel. You could choose any horizontal position to align them all on, but the one that costs the least fuel is horizontal position <code>2</code>:</p>
<ul>
<li>Move from <code>16</code> to <code>2</code>: <code>14</code> fuel</li>
<li>Move from <code>1</code> to <code>2</code>: <code>1</code> fuel</li>
<li>Move from <code>2</code> to <code>2</code>: <code>0</code> fuel</li>
<li>Move from <code>0</code> to <code>2</code>: <code>2</code> fuel</li>
<li>Move from <code>4</code> to <code>2</code>: <code>2</code> fuel</li>
<li>Move from <code>2</code> to <code>2</code>: <code>0</code> fuel</li>
<li>Move from <code>7</code> to <code>2</code>: <code>5</code> fuel</li>
<li>Move from <code>1</code> to <code>2</code>: <code>1</code> fuel</li>
<li>Move from <code>2</code> to <code>2</code>: <code>0</code> fuel</li>
<li>Move from <code>14</code> to <code>2</code>: <code>12</code> fuel</li>
</ul>
<p>This costs a total of <code><em>37</em></code> fuel. This is the cheapest possible outcome; more expensive outcomes include aligning at position <code>1</code> (<code>41</code> fuel), position <code>3</code> (<code>39</code> fuel), or position <code>10</code> (<code>71</code> fuel).</p>
<p>Determine the horizontal position that the crabs can align to using the least fuel possible. <em>How much fuel must they spend to align to that position?</em></p>


In [1]:
import pandas as pd 
import numpy as np

In [90]:
f = open("crabs.txt", "r")
raw = f.read()
raw[:100]

'1101,1,29,67,1102,0,1,65,1008,65,35,66,1005,66,28,1,67,65,20,4,0,1001,65,1,65,1106,0,8,99,35,67,101,'

In [95]:
crabs = np.fromstring(raw, sep = ',', dtype= int)
crabs[:10]

array([1101,    1,   29,   67, 1102,    0,    1,   65, 1008,   65])

In [96]:
crabs_s = np.sort(crabs)
crabs_s[:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [100]:
def busquedaBinaria(crabs, rango, consumo_extra= False):
    mitad = int(crabs[rango[0]:rango[1]].size/2)+rango[0]
    parte1 = crabs[rango[0]: mitad]
    parte2 = crabs[mitad: rango[1]]
    m1 = int(parte1.size/2)
    m2 = int(parte2.size/2)
    
    if not consumo_extra:
        v_parte1 = abs(crabs[:]-parte1[m1]).sum()
        v_parte2 = abs(crabs[:]-parte2[m2]).sum()
    else:
        v_parte1 = np.apply_along_axis(lambda x: (x*(x + 1))/2, 0,abs(crabs[:]-parte1[m1])).sum()
        v_parte2 = np.apply_along_axis(lambda x: (x*(x + 1))/2, 0,abs(crabs[:]-parte2[m1])).sum()
    
    if v_parte1<v_parte2:
        if parte1.size == 1:
            return v_parte1, m1+rango[0]
        else:
            return busquedaBinaria(crabs, [rango[0],mitad], consumo_extra)
    else:
        if parte2.size == 1:
            return v_parte2, m2+rango[0]
        else:
            return busquedaBinaria(crabs, [mitad, rango[1]], consumo_extra)

In [101]:
consumo , idx_posicion = busquedaBinaria(crabs_s, [0, len(crabs)])
consumo, idx_posicion

(364898, 498)

#### resultado 364898

<h2 id="part2">--- Part Two ---</h2><p>The crabs don't seem interested in your proposed solution. Perhaps you misunderstand crab engineering?</p>
<p>As it turns out, crab submarine engines <span title="This appears to be due to the modial interaction of magneto-reluctance and capacitive duractance.">don't burn fuel at a constant rate</span>. Instead, each change of 1 step in horizontal position costs 1 more unit of fuel than the last: the first step costs <code>1</code>, the second step costs <code>2</code>, the third step costs <code>3</code>, and so on.</p>
<p>As each crab moves, moving further becomes more expensive. This changes the best horizontal position to align them all on; in the example above, this becomes <code>5</code>:</p>
<ul>
<li>Move from <code>16</code> to <code>5</code>: <code>66</code> fuel</li>
<li>Move from <code>1</code> to <code>5</code>: <code>10</code> fuel</li>
<li>Move from <code>2</code> to <code>5</code>: <code>6</code> fuel</li>
<li>Move from <code>0</code> to <code>5</code>: <code>15</code> fuel</li>
<li>Move from <code>4</code> to <code>5</code>: <code>1</code> fuel</li>
<li>Move from <code>2</code> to <code>5</code>: <code>6</code> fuel</li>
<li>Move from <code>7</code> to <code>5</code>: <code>3</code> fuel</li>
<li>Move from <code>1</code> to <code>5</code>: <code>10</code> fuel</li>
<li>Move from <code>2</code> to <code>5</code>: <code>6</code> fuel</li>
<li>Move from <code>14</code> to <code>5</code>: <code>45</code> fuel</li>
</ul>
<p>This costs a total of <code><em>168</em></code> fuel. This is the new cheapest possible outcome; the old alignment position (<code>2</code>) now costs <code>206</code> fuel instead.</p>
<p>Determine the horizontal position that the crabs can align to using the least fuel possible so they can make you an escape route! <em>How much fuel must they spend to align to that position?</em></p>


In [103]:
consumo , idx_posicion = busquedaBinaria(crabs_s, [0, len(crabs)], True)
consumo, idx_posicion

(104149091.0, 593)

#### resultado 104149091